src/backend/distributed/planner/multi_planner.c

```diff
@@ -294,176 +333,235 @@ CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query
-		FuncExpr *funcExpr = makeNode(FuncExpr);
-		TargetEntry *targetEntry = NULL;
-		bool resjunkAttribute = true;
+		executorType = JobExecutorType(multiPlan);
+	}
 
-		funcExpr->funcretset = true;
+	switch (executorType)
+	{
+		case MULTI_EXECUTOR_REAL_TIME:
+		{
+			customScan->methods = &RealTimeCustomScanMethods;
+			break;
+		}
 
-		targetEntry = makeTargetEntry((Expr *) funcExpr, InvalidAttrNumber, NULL,
-									  resjunkAttribute);
+		case MULTI_EXECUTOR_TASK_TRACKER:
+		{
+			customScan->methods = &TaskTrackerCustomScanMethods;
+			break;
+		}
+
+		case MULTI_EXECUTOR_ROUTER:
+		{
+			customScan->methods = &RouterCustomScanMethods;
+			break;
+		}
 
-		fauxFunctionScan->scan.plan.targetlist =
-			lappend(fauxFunctionScan->scan.plan.targetlist,
-					targetEntry);
+		default:
+		{
+			customScan->methods = &InvalidCustomScanMethods;
+			break;
+		}
 	}
 
-	result->planTree = (Plan *) fauxFunctionScan;
+	multiPlanData = SerializeMultiPlan(multiPlan);
 
-	return result;
+	customScan->custom_private = list_make1(multiPlanData);
+	customScan->flags = CUSTOMPATH_SUPPORT_BACKWARD_SCAN;
+
+	/* check if we have a master query */
+	if (multiPlan->masterQuery)
+	{
+		finalPlan = FinalizeNonRouterPlan(localPlan, multiPlan, customScan);
+	}
+	else
+	{
+		finalPlan = FinalizeRouterPlan(localPlan, multiPlan, customScan);
+	}
+
+	return finalPlan;
 }
 
 
 /*
- * GetMultiPlanString returns either NULL, if the plan is not a distributed
- * one, or the string representing the distributed plan.
+ * FinalizeNonRouterPlan gets the distributed custom scan plan, and creates the
+ * final master select plan on the top of this distributed plan for real-time
+ * and task-tracker executors.
  */
-static char *
-GetMultiPlanString(PlannedStmt *result)
+static PlannedStmt *
+FinalizeNonRouterPlan(PlannedStmt *localPlan, MultiPlan *multiPlan,
+					  CustomScan *customScan)
 {
-	FunctionScan *fauxFunctionScan = NULL;
-	RangeTblFunction *fauxFunction = NULL;
-	FuncExpr *fauxFuncExpr = NULL;
-	Const *multiPlanData = NULL;
+	PlannedStmt *finalPlan = NULL;
 
-	if (!IsA(result->planTree, FunctionScan))
-	{
-		return NULL;
-	}
+	finalPlan = MasterNodeSelectPlan(multiPlan, customScan);
+	finalPlan->queryId = localPlan->queryId;
+	finalPlan->utilityStmt = localPlan->utilityStmt;
 
-	fauxFunctionScan = (FunctionScan *) result->planTree;
+	return finalPlan;
+}
 
-	if (list_length(fauxFunctionScan->functions) != 1)
-	{
-		return NULL;
-	}
 
-	fauxFunction = linitial(fauxFunctionScan->functions);
+/*
+ * FinalizeRouterPlan get the distributed router executor plan and wraps it with
+ * a proper target list.
+ */
+static PlannedStmt *
+FinalizeRouterPlan(PlannedStmt *localPlan, MultiPlan *multiPlan, CustomScan *customScan)
+{
+	PlannedStmt *routerPlan = NULL;
+	RangeTblEntry *customScanRangeTableEntry = NULL;
+	ListCell *targetEntryCell = NULL;
+	List *targetList = NIL;
+	List *columnNameList = NIL;
+
+	/* we have only one range table entry */
+	int customScanRangeTableIndex = 1;
```

Comment:
I updated the comment. Here, we will create only one range table entry for the router plan, so we know this by sure.